src/celeritas/optical/ScintillationData.hh

```diff
@@ -29,67 +34,154 @@ struct ScintillationComponent
+ * particle type.
+ */
+struct ParticleScintillationSpectrum
+{
+    GenericGridData yield_vector;
+    ItemRange<ScintillationComponent> components;
+
+    //! Whether all data are assigned and valid
+    explicit CELER_FUNCTION operator bool() const
+    {
+        return static_cast<bool>(yield_vector);
+    }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Data characterizing the scintillation spectrum for all particles and
+ * materials.
+ *
+ * Sampling using material-only data or particle- and material-dependent data
+ * are mutually exclusive. Therefore, either \c materials or \c particles are
+ * loaded at the beginning of the simulation, but *never* both at the same
+ * time. The \c scintillation_by_particle() function can be used to check that.
+ *
+ * - \c matid_to_optmatid returns an \c OpticalMaterialId given a
+ *   \c MaterialId
+ * - \c pid_to_scintpid returns a \c ScintillationParticleId given a
+ *   \c ParticleId .
+ * - \c resolution_scale is indexed by \c OpticalMaterialId .
+ * - \c materials stores material-only scintillation data. Indexed by
+ *   \c OpticalMaterialId
+ * - \c particles stores scintillation data for each particle type for each
+ *   material, being a grid of size `num_particles * num_materials`. Therefore
+ *   it is indexed by \c ParticleScintSpectrumId , which combines
+ *   \c ScintillationParticleId and \c OpticalMaterialId . Use the
+ *   \c spectrum_index() function to retrieve the correct index.
  */
 template<Ownership W, MemSpace M>
 struct ScintillationData
 {
     template<class T>
     using Items = Collection<T, W, M>;
-    template<class T>
-    using OpticalMaterialItems = Collection<T, W, M, OpticalMaterialId>;
+    using MaterialItems
+        = Collection<MaterialScintillationSpectrum, W, M, OpticalMaterialId>;
+    using ParticleItems
+        = Collection<ParticleScintillationSpectrum, W, M, ParticleScintSpectrumId>;
 
     //// MEMBER DATA ////
 
+    //! Index between OpticalMaterialId and MaterialId
+    Collection<OpticalMaterialId, W, M, MaterialId> matid_to_optmatid;
+    //! Index between ScintillationParticleId and ParticleId
+    Collection<ScintillationParticleId, W, M, ParticleId> pid_to_scintpid;
+
+    //! Resolution scale for each material
+    Collection<real_type, W, M, OpticalMaterialId> resolution_scale;
+
+    //! Material-only scintillation spectrum data
+    MaterialItems materials;  //!< [OpticalMaterialId]
+
+    //! Particle and material scintillation spectrum data
+    ParticleItems particles;  //!< [ParticleScintSpectrumId]
+
+    //! Backend storage for ParticleScintillationSpectrum::yield_vector
+    Items<real_type> grid_data;
+
+    //! Components for either material or particle items
     Items<ScintillationComponent> components;
-    OpticalMaterialItems<ScintillationSpectrum> spectra;
+
+    size_type num_opt_materials{};
+    size_type num_opt_particles{};
```

Comment:
The `materials` yes. The `particles` is not, since `particles.size() == num_opt_materials * num_opt_particles`.
But your point is valid, since I guess we also won't really need those during the main loop, as we always convert material or particle id into their respective scintillation id and access the data directly from those ids.